src/coord/rmr/redise.c

```diff
@@ -111,82 +134,199 @@ MRClusterTopology *RedisEnterprise_ParseTopology(RedisModuleCtx *ctx, RedisModul
+
+      } else if (AC_AdvanceIfMatch(&ac, "MASTER")) {
+        is_master = true;
+      } else {
+        break;
       }
     }
-    if (MREndpoint_Parse(addr, &sh.node.endpoint) != REDIS_OK) {
-      ERROR_BADVAL("ADDR", addr);
-      goto error;
+
+    // We don't care for replicas using this command anymore
+    if (!is_master) {
+      RLShard_Free(sh);
+      sh = NULL;
+      continue;
     }
-    /* Optional MASTER */
-    if (!AC_AdvanceIfMatch(&ac, "MASTER")) {
-      // We don't care for replicas using this command anymore
-      MREndpoint_Free(&sh.node.endpoint);
+
+    // Ignore shards with no slot ranges (like replicas)
+    if (array_len(sh->slotRanges) == 0) {
+      RLShard_Free(sh);
+      sh = NULL;
       continue;
     }
-    // All good. Finish up the node
-    sh.node.id = rm_strdup(sh.node.id);  // Take ownership of the string
-    if (unixSock) {
-      sh.node.endpoint.unixSock = rm_strdup(unixSock);
+
+    dictEntry *entry = dictAddOrFind(shards, shardIDStr);
+    if (!dictGetVal(entry)) {
+      // New shard
+      // Verify mandatory arguments on first appearance
+      if (!sh->node.endpoint.host) {
+        ERROR_MISSING("ADDR");
+        goto error;
+      }
+      // Move ownership of parsed shard into dict
+      dictSetVal(shards, entry, sh);
+      sh = NULL;
+    } else {
+      // Re-appearance of shard ID
+      // We verify that the endpoint is the same
+      // We also verify that slot range is different from previous ones
+      RLShard *existing_shard = dictGetVal(entry);
+
+      // Verify endpoint, if currently specified
+      if (sh->node.endpoint.host) {
+
+        if (strcmp(sh->node.endpoint.host, existing_shard->node.endpoint.host)) {
+          ERROR_FMT("Conflicting ADDR for shard `%s`", sh->node.id);
+          goto error;
+        }
+        if ((sh->node.endpoint.password && !existing_shard->node.endpoint.password) ||
+            (!sh->node.endpoint.password && existing_shard->node.endpoint.password) ||
+            (sh->node.endpoint.password && existing_shard->node.endpoint.password && strcmp(sh->node.endpoint.password, existing_shard->node.endpoint.password))) {
+          ERROR_FMT("Conflicting ADDR for shard `%s`", sh->node.id);
+          goto error;
+        }
+        if (sh->node.endpoint.port != existing_shard->node.endpoint.port) {
+          ERROR_FMT("Conflicting ADDR for shard `%s`", sh->node.id);
+          goto error;
+        }
+      }
+      if (sh->node.endpoint.unixSock) {
+        if (!existing_shard->node.endpoint.unixSock || strcmp(sh->node.endpoint.unixSock, existing_shard->node.endpoint.unixSock)) {
+          ERROR_FMT("Conflicting UNIXADDR for shard `%s`", sh->node.id);
+          goto error;
+        }
+      }
+      RS_ASSERT(array_len(sh->slotRanges) == 1);
+      // Verify slot range starts past existing ones
+      if (array_tail(existing_shard->slotRanges).end + 1 >= sh->slotRanges[0].start) {
```

Comment:
Should we always assume that slot ranges will come in increasing order? Is 
```
RANGES 2 SHARD 1 SLOTRANGE 10 20 [...] SHARD 1 SLOTRANGE 1 9 [...]
```
an error?

Reply:
Yes, this is agreed upon